crates/lib/src/execution_data.rs

```diff
@@ -1,58 +1,121 @@
+use enum_as_inner::EnumAsInner;
+use num::complex::{Complex32, Complex64};
 use std::collections::HashMap;
-use std::convert::TryFrom;
-use std::num::{ParseIntError, TryFromIntError};
+use std::num::ParseIntError;
 use std::str::FromStr;
 use std::time::Duration;
 
-use qcs_api_client_grpc::models::controller::ReadoutValues;
+use ndarray::prelude::*;
+
+use qcs_api_client_grpc::models::controller::{readout_values, ReadoutValues};
 use quil_rs::instruction::MemoryReference;
 
 use crate::RegisterData;
 
-/// The result of executing an [`Executable`](crate::Executable) via
-/// [`Executable::execute_on_qvm`](crate::Executable::execute_on_qvm).
+/// The result of executing an [`Executable`](crate::Executable)
 #[derive(Debug, Clone, PartialEq)]
-pub struct Qvm {
+pub struct ExecutionData {
     /// The readout data that was read from the [`Executable`](crate::Executable).
     /// Key is the name of the register, value is the data of the register after execution.
-    pub registers: HashMap<Box<str>, RegisterData>,
+    pub readout_data: ReadoutMap,
     /// The time it took to execute the program on the QPU, not including any network or queueing
     /// time. If paying for on-demand execution, this is the amount you will be billed for.
     ///
     /// This will always be `None` for QVM execution.
     pub duration: Option<Duration>,
 }
 
+/// An enum representing every possible readout type
+#[derive(Clone, Copy, Debug, EnumAsInner, PartialEq)]
+pub enum ReadoutValue {
+    /// TODO
+    I8(i8),
+    /// TODO
+    I16(i16),
+    /// TODO
+    I32(i32),
+    /// TODO
+    F64(f64),
+    /// TODO
+    Complex32(Complex32),
+    /// TODO
+    Complex64(Complex64),
+}
+
+/// A matrix where rows are the values of a memory index across all shots, and columns are values
+/// for all memory indexes of a given shot.
+pub type RegisterMatrix = Array2<Option<ReadoutValue>>;
+
 #[derive(Debug, Clone, PartialEq)]
 #[repr(transparent)]
-pub struct ReadoutMap(HashMap<MemoryReference, ReadoutValues>);
+pub struct ReadoutMap(HashMap<String, RegisterMatrix>);
 
 impl ReadoutMap {
-    /// Given a known readout field name and index, return the result's [`ReadoutValues`], if any.
-    pub fn get_readout_values(&self, field: String, index: u64) -> Option<ReadoutValues> {
-        let readout_values = self.0.get(&MemoryReference { name: field, index })?;
+    /// Returns the individual value for the given memory index and shot number, or None, if it
+    /// doesn't exist
+    pub fn get_value(
+        &self,
+        register_name: &str,
+        index: usize,
+        shot_num: usize,
+    ) -> Option<ReadoutValue> {
+        let register = self.0.get(register_name);
+        if let Some(matrix) = register {
+            if index > matrix.nrows() || shot_num > matrix.ncols() {
```

Comment:
Is there an off by one error here?
```suggestion
            if index >= matrix.nrows() || shot_num >= matrix.ncols() {
```

I think the Rusty thing to do is something more like

```rust
self.0.get(register_name)
    .and_then(|matrix| matrix.get((index, shot_num)))
    .copied()
```

No manual bounds checks, no accidental panics.

Reply:
Thank you! I could feel that there was a more Rusty thing to do here and I knew you would call it out. Applied your suggestion, just had to throw a `?` at the end as well.